lib/gooddata/lcm/actions/update_metric_formats.rb

```diff
@@ -0,0 +1,159 @@
+          return nil if metrics_hash.empty?
+
+          expected_keys = %w[tag client_id format]
+          unless expected_keys.map(&:to_sym).all? { |s| metrics_hash.first.key? s }
+            GoodData.logger.warn("The input metric data is incorrect, expecting at least the following fields: #{expected_keys}")
+            return nil
+          end
+          metrics_hash
+        end
+
+        def validate_input_source(input_source)
+          type = input_source[:type] if input_source&.dig(:type)
+          metric_format = input_source[:metric_format] if input_source&.dig(:metric_format)
+          return nil if type.blank? || metric_format.blank?
+
+          modified_input_source = input_source
+          case type
+          when 'ads', 'redshift', 'snowflake', 'bigquery', 'postgresql', 'mssql'
+            return nil if metric_format[:query].blank?
+
+            modified_input_source[:query] = metric_format[:query]
+            return modified_input_source
+          when 's3'
+            return nil if metric_format[:file].blank?
+
+            if modified_input_source.key?(:key)
+              modified_input_source[:key] = metric_format[:file]
+            else
+              modified_input_source[:file] = metric_format[:file]
+            end
+            return modified_input_source
+          when 'blobStorage'
+            return nil if metric_format[:file].blank?
+
+            modified_input_source[:file] = metric_format[:file]
+            return modified_input_source
+          when 'staging'
+            return nil if metric_format[:file].blank?
+
+            modified_input_source[:path] = metric_format[:file]
+            return modified_input_source
+          when 'web'
+            return nil if metric_format[:url].blank?
+
+            modified_input_source[:url] = metric_format[:url]
+            return modified_input_source
+          else
+            return nil
+          end
+        end
+
+        def get_clients_metrics(metric_data)
+          return {} if metric_data.nil?
+
+          metric_groups = {}
+          clients = metric_data.map { |row| row[:client_id] }.uniq
+          clients.each do |client|
+            next if client.blank?
+
+            formats = {}
+            metric_data.select { |row| row[:client_id] == client && row[:tag].present? && row[:format].present? }.each { |row| formats[row[:tag]] = row[:format] }
+            metric_groups[client] ||= formats
+          end
+          metric_groups
+        end
+
+        def call(params)
+          result = []
+          data = load_metric_data(params)
+          return result if data.nil?
+
+          metric_group = get_clients_metrics(data)
+          return result if metric_group.empty?
+
+          updated_clients = params.synchronize.map { |segment| segment.to.map { |client| client[:client_id] } }.flatten.uniq
+          data_product = params.data_product
+          data_product_clients = data_product.clients
```

Comment:
Add execution log:
- begin and end updating metrics formats.
- begin and end of each project
-

Reply:
Added logs, thanks